consent/manager_sql.go

```diff
@@ -287,7 +287,38 @@ func (m *SQLManager) HandleConsentRequest(ctx context.Context, challenge string,
 		strings.Join(sqlParamsConsentRequestHandled, ", "),
 		":"+strings.Join(sqlParamsConsentRequestHandled, ", :"),
 	), d); err != nil {
-		return nil, sqlcon.HandleError(err)
+		err = sqlcon.HandleError(err)
+		if errors.Cause(err) == sqlcon.ErrUniqueViolation {
+			return m.ReplaceUnusedConsentRequest(ctx, challenge, r)
+		}
+		return nil, err
+	}
+
+	return m.GetConsentRequest(ctx, challenge)
+}
+
+func (m *SQLManager) ReplaceUnusedConsentRequest(ctx context.Context, challenge string, r *HandledConsentRequest) (*ConsentRequest, error) {
+	d, err := newSQLHandledConsentRequest(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := m.DB.NamedQueryContext(ctx, fmt.Sprintf(`
```

Comment:
Query implements a scanner for things like `SELECT`, not an executor, see [query](https://golang.org/pkg/database/sql/#DB.Query) versus [exec](https://golang.org/pkg/database/sql/#DB.Exec) in godoc. If you update this to `NamedExecContext` be aware that it won't return `sqlcon.ErrNoRows` if no rows are affected. Instead, you need to check the first argument, probably something like `results.ReowsAffected()`. If it's 0, the update failed.

Reply:
That makes the logic a lot cleaner, actually. I return `m.GetConsentRequest` regardless of how many rows were affected. I just needed to make sure that if an error was returned, then take the appropriate action. (I thought 0 rows affected would apply here)